### ISSUE TITLE
Pointer for end-users to the new location of the Qt reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2137]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2137|
|Type|task|
|Created|2006-09-30 18:08:03Z|

Since qtreactor will not be present in Twisted 2.5, but will be available as a separate download, we should have some information for people trying to use it about where they can get the software to make it work.

This should be mentioned in the release notes, of course, but it may also be productive to have ```twistd --reactor qt``` spit out an URL.

<details><summary>Searchable metadata</summary>

```
trac-id__2137 2137
type__task task
reporter__exarkun exarkun
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1159639683000000 1159639683000000
changetime__1166846651000000 1166846651000000
version__None None
owner__ 
cc__glyph cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I thought about having twisted.internet.qtreactor raise an exception or otherwise print a message, which could have been done in the [#2130](https://github.com/twisted/twisted/issues/2130) branch, but this would have made it harder to have a plugin for qtreactor.  I think we may just have to special case "qt" in the reactor selection code for a while.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [18345]) Merge remove-qtreactor-2130

Author: exarkun
Reviewer: jml
Fixes [#2130](https://github.com/twisted/twisted/issues/2130)
Refs [#2137](https://github.com/twisted/twisted/issues/2137)

This removes support for Qt event loop integration from Twisted, due to
licensing concerns.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to PenguinOfDoom|
|-|-|

Thank you for working on this.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|PenguinOfDoom set owner to @exarkun|
|-|-|

Code is ready for review in qtreactor-stub-2137. Wiki page is http://twistedmatrix.com/trac/wiki/QTReactor.

This ticket will not be resolved until the two files (qtreactor.py and qtdemo.py) grow an appropriate license.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Haven't done a complete review yet, but:

  * QTStubProcessProtocol needs a docstring
  * test_qtStub needs a docstring

Any reason not to use t.i.u.getProcessOutputAndValue?  It seems to do the same thing as much of this code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to PenguinOfDoom|
|-|-|

I guess something doing:

```
from twisted.internet import qtreactor
qtreactor.install()
```

should keep working as well.

Other stuff:

  * don't use \ line continuations
  * ```wiki_url``` should be ```wikiURL```
  * Won't ```import qtreactor``` result in Qt being imported into every Twisted application run on a system with qtreactor installed?  Hopefully most people won't have it installed, but we probably shouldn't do this to people who do.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|PenguinOfDoom commented|
|-|-|

I addressed these concerns.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to PenguinOfDoom|
|-|-|

Docstrings, getProcessOutputAndValue, etc.

Add a deprecation warning to the install method in qtreactor_stub so people stop using it.  Stick a "qt3" plugin on the wiki page, as discussed on irc.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|PenguinOfDoom set owner to @exarkun|
|-|-|

It's better now. I didn't rename the reactor itself to QT3Reactor or something. I don't think that's necessary.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to PenguinOfDoom|
|-|-|

Underscores are used as a replacement for dots in plugin names, due to scoping issues.  ```twisted.plugin.qt_stub``` should be ```twisted.plugin.twisted_qtstub```.

Docstrings still missing:
 * ```twisted.plugin.qt_stub```
 * ```twisted.plugin.qt_stub.QTStub```
 * ```twisted.plugin.qt_stub.QTStub.install```
 * ```app_qtstub.py``` - since this is a test fixture, only the file needs a docstring, but it's impossible to tell what it's for in isolation.  when I first looked at it I mistakenly thought that it was supposed to be imported...

```test_qtStub``` ought to be passing sys.executable as the 0th arg as well as the binary; no reason to confuse python unduly about its installation location.
```app_qtstub.py``` should probably be renamed to ```app-qtstub.py``` or similar to avoid the unfortunate possibility that it will be imported by something - like, let's say, a future version of pydoctor.  Also in this file, ```raise ImportError('qtreactor does not exist!')```, not ```raise ImportError, ...```

Obviously not a requirement of the merge here, but it would be nice if the docstring in the qtreactor module in the tarball on the wiki page didn't refer to using twisted.internet.qtreactor to install, and instead referred to itself (an unqualified import, apparently).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @glyph|
|-|-|

Fixed those issues, I think.  Also did some other stuff I couldn't resist doing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

As of now, looks good to merge.  Do it.

However, http://twistedmatrix.com/users/pahan/qt_file/qtdemo.py - referenced from the wiki page - is a dead link.  I'm just deleting the link for now, but perhaps someone will want to put it back later.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [19009]) Merge qtreactor-stub-2137

Author: PenguinOfDoom
Reviewer: glyph
Fixes [#2137](https://github.com/twisted/twisted/issues/2137)

This adds a reactor plugin which will attempt to import qtreactor and
warn that the name qt is now deprecated.  If qtreactor cannot be
imported it will point to the location from which it can be retrieved.

This also adds a similar backwards-compatibility stub for
twisted.internet.qtreactor for applications which install it themselves
rather than using twistd.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
